### ISSUE TITLE
Repeated creation and removal of table with index causes significant slowdown of scdata

### ISSUE THREAD

**Issue by henriksjostrom** [Starcounter/Starcounter#3950]
When repeatedly creating a table with an index and some data and removing it the scdata process becomes increasingly slow. 

Repeated execution of the testcase with the same scdata process causes further slowdowns. Restarting scdata resets the speed to the original.

I have created a testcase for this [here](https://git.starcounter.org/BlueStar/level0/blob/scdata_slowdown/src/tests/slowdown/slowdown.cpp)

[here](https://teamcity.starcounter.org/viewLog.html?buildId=13691&buildTypeId=Starcounter_LinuxNextDailyRelease&tab=buildResultsDiv) is a failed personal build showing the problem on Linux but the problem exists both on Linux and windows.

Not creating an index causes each iteration to not take longer time. It seems likely that this somehow relates to indexes in the kernel.

@bigwad, @k-rus

**Comment by k-rus** [comment]
I feel it is unexpected behaviour. Let's see what @bigwad says.

Just wonder if you don't delete table, but create another one, e.g., by suffixing table name with test id, do you observe the same picture?

I guess it is necessary to have a work around in the test framework.

**Comment by henriksjostrom** [comment]
@k-rus It appears that the problem persists if I do not drop the tables.

**Comment by henriksjostrom** [comment]
For the record the increased delay appears to be roughly ∝x^2 where x is the number of times the table has been added/dropped. Preferably it should be constant.

**Comment by alexanderstc** [comment]
> Preferably it should be constant.

Looky, such a cute fat little troll!

**Comment by bigwad** [comment]
The problem is around setspec and code generation in QP (not related to scllvm).

First, scllvm maintaines a cache of compiled code. It matches cpp source with compiled code by the hash of cpp source. So, while QP gives scllvm the same input - it returns compiled module from the cache.
Among other variable parts, this cpp source contains setspec in the string form. So, if setspec changes, scllvm can't find corresponding module in the cache and compiles a new one, that is slow.

To the problem - looks like when you drop a table and then create it again, metalayer is able reuse setspec for the new table. Sometimes. But sometimes not. In latter case you see performance drop due. I don't know if it's a problem. Probably not, as it's unlikely be a real life scenario, when you recreate a table with exactly the same layout.

And finally, what odds I've found during this investigation.
Don't see why setspec is hardcoded into generated code when QP scans for all records in a table. If you want to hardcode something - it would make some sense to hardcode encoded setspec at least. Even better would be to avoid runtime index range composition and hardcode preformatted index range. But the best I think is to get rid of generated code in such case, when you know index structure in advance.

**Comment by k-rus** [comment]
@bigwad Thank you for the investigation!

**Comment by bigwad** [comment]
> metalayer is able to reuse setspec for the new table. Sometimes. But sometimes not.

I was wrong about it. setspec never gets reused. When you see good performance- it just means that scllvm  cache still holds compiled code corresponing for this setspec. As setspec algorithm is determenistic, it regenerate setspecs in the same order on every recreated database.
So if you cleanup the cache - you'll see bad performance from very first run. Then if you restart your test (but not cleanup the cache) - then you'll be getting good performance until your tests advances to setspec that hasn't been generated on the previous run.

**Comment by k-rus** [comment]
> First, scllvm maintaines a cache of compiled code. It matches cpp source with compiled code by the hash of cpp source. So, while QP gives scllvm the same input - it returns compiled module from the cache.

I will elaborate about it a little bit. With new scllvm the compiled functions are stored on disk. Thus when the test is restarted it will reuse compiled functions from previous run.

> I was wrong about it. setspec never gets reused. When you see good performance- it just means that scllvm cache still holds compiled code corresponing for this setspec. As setspec algorithm is determenistic, it regenerate setspecs in the same order on every recreated database.
> So if you cleanup the cache and the database - you'll see bad performance from very first run. Then if you restart your test on the fresh database (but do not cleanup the cache) - then you'll be getting good performance until your tests advances to setspec that hasn't been generated on the previous run.

Thank you for confirming this. It is expected to work in this way.

**Comment by alexanderstc** [comment]
Should I pass setspec as an argument to the compiled function?

**Comment by k-rus** [comment]
> Should I pass setspec as an argument to the compiled function?

Nope, different setspec means different table and those different query plan.

**Comment by alexanderstc** [comment]
> different setspec means different table and those different query plan.

If the query plan is really different, the generated query code will be different as well (not only setspec).
If I pass setspec as an argument, scllvm will be able to cache the code for those queries which have exactly the same code (meaning the same query plans).

**Comment by k-rus** [comment]
> If the query plan is really different, the generated query code will be different as well (not only setspec).
> If I pass setspec as an argument, scllvm will be able to cache the code for those queries which have exactly the same code (meaning the same query plans).

If set spec is different, then layout id is very likely to be different as well. Actually there will be cases when code gen is overkill.

**Comment by alexanderstc** [comment]
OK, I cancel my patches.

**Comment by dan31** [comment]
Interesting investigation! But what is the outcome?

**Comment by k-rus** [comment]
> Interesting investigation! But what is the outcome?

@henriksjostrom did some improvements to the really drop table and made new sets of experiments. @henriksjostrom can you share your last graph?

**Comment by henriksjostrom** [comment]
Yes

TL:DR (or abstract in academican) Creating a table, filling it with data is constant. If an index is created or a query executed the time appears to increase linearly. 

First of all here is the initial graph that lead me to state ∝x^2
![image](https://cloud.githubusercontent.com/assets/2777441/21048782/576ef78a-be11-11e6-8ba4-b7f7908c57b0.png)

After I rewrote the really_drop_table function and some set of commits were done in pnext repeating the experiment under some conditions give the following results:

When running neither a query nor an index the time increases linearly but resets itself periodicly becoming effectively constant in time.
![image 4](https://cloud.githubusercontent.com/assets/2777441/21048809/84e50948-be11-11e6-9b5d-7f9d595a9f4c.png)

Running a query and no index creation on each iteration
![image 3](https://cloud.githubusercontent.com/assets/2777441/21048811/89ce4bb8-be11-11e6-852d-ef106d6726c4.png)
Running a query and having an index, the jump in timing was a surprise so I reran the experiment twice more and it didn't appear so it can probably be disregarded.
![image 2](https://cloud.githubusercontent.com/assets/2777441/21048812/89e606b8-be11-11e6-8f15-951cddf2eaf0.png)
No query but creating an index
![image 1](https://cloud.githubusercontent.com/assets/2777441/21048813/89f13d6c-be11-11e6-93bb-f222b5ca2712.png)
Between each execution the database was restarted and all files removed. The executions were all done on my computer not on a server and not as the only running application so exact numbers should be taken with a grain of salt.

[This is the code](https://git.starcounter.org/BlueStar/level0/blob/1904a77b6ffecd7cef2af042e48e5b29d37c4774/src/tests/slowdown/slowdown.cpp) that was used to generate the dataset

**Comment by k-rus** [comment]
cc @dan31

**Comment by dan31** [comment]
@henriksjostrom awesome graphs! Why not in LaTeX though? 😃 
Maybe we could run profiler on top of that to see what increases over runs, at our next meeting.

**Comment by henriksjostrom** [comment]
@dan31 While as for any true physicist LaTeX should naturally have been my first choice. Alas the setup time required led me to prioritize a faster solution. I hope to one day earn your forgiveness.

Yes, I would like to learn how to attach a debugger to the scdata process rather than the test process. It sounds like it could be useful to know. CPU time spent in the test process is roughly equal between 2 runs where the scdata process hasn't been restarted in between (while actual execution time is much slower on the 2nd run)

**Comment by dan31** [comment]
The lesson we may derive from the issue: it is relevant to run perf/load tests also in Debug (e.g., it allows to identify runtime assertions of STL containers). This is what Henrik did. As a result of run being done in Debug, this code in VTrie implementation has been always triggered:

![2016-12-16](https://cloud.githubusercontent.com/assets/7162535/21288239/d26738d6-c47d-11e6-8ac3-22db3865c3b9.png)

As more items are added during the test run, more bytes constitute VTrie. Debug build faithfully forms ```vtrie_bump``` even though it is not used anywhere else (and thus a subject of optimising elimination). As a result, we see the behavior from @henriksjostrom graphs. This makes Debug run be a [Shlemiel the painter’s algorithm](https://www.joelonsoftware.com/2001/12/11/back-to-basics/).

In Release, this function is not used (see preprocessor). Our guess: the Debug print of vtrie into non-used stringstream should be removed. Perhaps what was meant is not ```DEBUG```, but rather ```KERNEL_DEBUG``` to be able to check vtrie by eyes on breakpoint. Or it might have been simply forgotten to be removed from there. Either way, this part should no longer affect Debug runs. Assigning to @bigwad.

As @henriksjostrom, @Thanh-Truong and @andan342 are now supplied with the knowledge on using out-of-the-box VS profiler, next time such issue appears they would be able to draw needed conclusions on their own, without treating the other parts of Starcounter as a black-box.

**Comment by dan31** [comment]
>  Don't really see why there is a need to run performance test in debug mode. 

Don't think "performance", take just "tests". As we see, it actually gives some unusual discoveries. Thanks for fixing!
